### PR TITLE
Convert ActionClient to use C++ classes

### PR DESCRIPTION
**Pull request by ahcorde**
Related to https://github.com/ros2/rclpy/issues/665

This converts the ActionClient functions to a C++ class. 

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
Tests will be green when this other PR https://github.com/ros2/rclpy/pull/756 is in

File: rclpy/src/rclpy/action_client.cpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    action_name,
+    &action_client_ops);
+  if (RCL_RET_ACTION_NAME_INVALID == ret) {
+    std::string error_text{"Failed to create action client due to invalid topic name '"};
+    error_text += action_name;
+    error_text += "' : ";
+    error_text += rcl_get_error_string().str;
+    rcl_reset_error();
+    throw py::value_error(error_text);
+  } else if (RCL_RET_OK != ret) {
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:100]: @ahcorde nit: this `else` clause is redundant.

File: rclpy/src/rclpy/action_client.cpp [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    error_text += rcl_get_error_string().str;
+    rcl_reset_error();
+    throw py::value_error(error_text);
+  }
+}
+
+#define TAKE_SERVICE_RESPONSE(Type) \
+  destroy_ros_message_signature * destroy_ros_message = NULL; \
+  /* taken_msg is always destroyed in this function */ \
+  void * taken_msg = rclpy_create_from_py(pymsg_type.ptr(), &destroy_ros_message); \
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:111]: @ahcorde nit: you may use `rclpy::create_from_py` instead.

File: rclpy/src/rclpy/action_client.cpp [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+  if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret || RCL_RET_ACTION_SERVER_TAKE_FAILED == ret) { \
+    pytuple[0] = py::none(); \
+    pytuple[1] = py::none(); \
+    return pytuple; \
+  } else if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to take " #Type); \
+  } \
+  pytuple[0] = py::int_(sequence); \
+  pytuple[1] = py::reinterpret_steal<py::object>( \
+    rclpy_convert_to_py(taken_msg_ptr.release(), pymsg_type.ptr())); \
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:132]: @ahcorde nit: you may use `rclpy::convert_to_py`

File: rclpy/src/rclpy/action_client.cpp [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+  rmw_request_id_t header; \
+  rcl_ret_t ret = rcl_action_take_ ## Type ## _response( \
+    rcl_action_client_.get(), &header, taken_msg); \
+  int64_t sequence = header.sequence_number; \
+  /* Create the tuple to return */ \
+  py::tuple pytuple(2); \
+  if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret || RCL_RET_ACTION_SERVER_TAKE_FAILED == ret) { \
+    pytuple[0] = py::none(); \
+    pytuple[1] = py::none(); \
+    return pytuple; \
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:126]: @ahcorde fyi: there's `py::make_tuple()` too.

File: rclpy/src/rclpy/action_client.cpp [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    throw rclpy::RCLError(error_text);
+  }
+
+  py::tuple result_tuple(5);
+  result_tuple[0] = py::int_(num_subscriptions);
+  result_tuple[1] = py::int_(num_guard_conditions);
+  result_tuple[2] = py::int_(num_timers);
+  result_tuple[3] = py::int_(num_clients);
+  result_tuple[4] = py::int_(num_services);
+  return result_tuple;
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:244]: @ahcorde fyi: you can compress this a bit using `py::make_tuple`

File: rclpy/src/rclpy/action_client.cpp [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  return is_available;
+}
+
+void
+ActionClient::wait_set_add(py::capsule pywait_set)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+
+  rcl_ret_t ret;
+  ret = rcl_action_wait_set_add_action_client(wait_set, rcl_action_client_.get(), NULL, NULL);
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:267]: @ahcorde nit: collapse the last two lines into one.

File: rclpy/src/rclpy/action_client.cpp [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+  result_tuple[0] = py::int_(num_subscriptions);
+  result_tuple[1] = py::int_(num_guard_conditions);
+  result_tuple[2] = py::int_(num_timers);
+  result_tuple[3] = py::int_(num_clients);
+  result_tuple[4] = py::int_(num_services);
+  return result_tuple;
+}
+
+bool
+ActionClient::is_available()
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:248]: @ahcorde nit: `is_action_server_available()` reads better.

File: rclpy/src/rclpy/action_client.cpp [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+  rcl_ret_t ret = rcl_action_server_is_available(
+    node, rcl_action_client_.get(), &is_available);
+  if (RCL_RET_OK != ret) {
+    throw rclpy::RCLError("Failed to check if action server is available");
+  }
+  return is_available;
+}
+
+void
+ActionClient::wait_set_add(py::capsule pywait_set)
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:262]: @ahcorde nit: would you consider `add_to_waitset` instead?

File: rclpy/src/rclpy/action_client.cpp [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+  rcl_ret_t ret;
+  ret = rcl_action_wait_set_add_action_client(wait_set, rcl_action_client_.get(), NULL, NULL);
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to add 'rcl_action_client_t' to wait set"};
+    throw rclpy::RCLError(error_text);
+  }
+}
+
+py::tuple
+ActionClient::wait_set_is_ready(py::capsule pywait_set)
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:275]: @ahcorde nit: would you consider `is_ready`?

File: rclpy/src/rclpy/action_client.hpp [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+  /// Take an action goal response.
+  /**
+   * Raises AttributeError if there is an issue parsing the pygoal_response_type.
+   * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
+   *
+   * \param[in] pygoal_response_type An instance of the response message type to take.
+   * \return 2-tuple (sequence number, received response), or
+   * \return 2-tuple (None, None) if there is no response, or
+   * \return NULL if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:70]: @ahcorde NULL ? Same below.

File: rclpy/src/rclpy/action_client.hpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    const rmw_qos_profile_t & goal_service_qos,
+    const rmw_qos_profile_t & result_service_qos,
+    const rmw_qos_profile_t & cancel_service_qos,
+    const rmw_qos_profile_t & feedback_topic_qos,
+    const rmw_qos_profile_t & status_topic_qos);
+
+  /// Take an action goal response.
+  /**
+   * Raises AttributeError if there is an issue parsing the pygoal_response_type.
+   * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:65]: @ahcorde is `RCLError` missing here?

File: rclpy/src/rclpy/action_client.hpp [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+  {
+    return rcl_action_client_.get();
+  }
+
+  /// Get rcl_client_t pointer
+  std::shared_ptr<rcl_action_client_t>
+  get_rcl_shared_ptr()
+  {
+    return rcl_action_client_;
+  }
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:223]: @ahcorde unused?

**Review by hidmic — COMMENTED** [review]
First pass !

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14297/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9086/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11983/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14401/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14315/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9104/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12001/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14420/)

File: rclpy/src/rclpy/action_client.cpp [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+  rcl_ret_t ret = rcl_action_take_ ## Type ## _response( \
+    rcl_action_client_.get(), &header, taken_msg.get()); \
+  int64_t sequence = header.sequence_number; \
+  /* Create the tuple to return */ \
+  if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret || RCL_RET_ACTION_SERVER_TAKE_FAILED == ret) { \
+    return py::make_tuple(py::none(), py::none()); \
+  } else if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to take " #Type); \
+  } \
+  return py::make_tuple(py::int_(sequence), convert_to_py(taken_msg.get(), pymsg_type)); \
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:123]: @ahcorde nit^N: I think you can just pass `sequence` and `pybind11` will do the magic:
```suggestion
  return py::make_tuple(sequence, convert_to_py(taken_msg.get(), pymsg_type)); \
```

File: rclpy/src/rclpy/action_client.cpp [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+    &num_clients,
+    &num_services);
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to get number of entities for 'rcl_action_client_t'"};
+    throw rclpy::RCLError(error_text);
+  }
+
+  return py::make_tuple(
+    py::int_(num_subscriptions), py::int_(num_guard_conditions), py::int_(num_timers),
+    py::int_(num_clients), py::int_(num_services));
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:224]: @ahcorde I'd expect:
```suggestion
  return py::make_tuple(
    num_subscriptions, num_guard_conditions, num_timers,
    num_clients, num_services);
```
to work.

File: rclpy/src/rclpy/action_client.hpp [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+namespace py = pybind11;
+
+namespace rclpy
+{
+class ActionClient : public Destroyable, public std::enable_shared_from_this<ActionClient>
+{
+public:
+  /// Create an action client.
+  /**
+   * This function will create an action client for the given action name.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:36]: @ahcorde nit: I wonder if this should be rephrased into (and put as) class level documentation.

File: rclpy/src/rclpy/action_client.hpp [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+  /// Take an action cancel response.
+  /**
+   * Raises AttributeError if there is an issue parsing the pycancel_response_type.
+   * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
+   * Raises RCLError if an error occurs in rcl
+   *
+   * \param[in] pycancel_response_type An instance of the response message type to take.
+   * \return 2-tuple (sequence number, received response), or
+   * \return 2-tuple (None, None) if there is no response, or
+   * \return (None, None) if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:97]: @ahcorde won't it `raise` if there's a failure?

File: rclpy/src/rclpy/action_client.hpp [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+  /// Take an action goal response.
+  /**
+   * Raises AttributeError if there is an issue parsing the pygoal_response_type.
+   * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
+   * Raises RCLError if an error occurs in rcl
+   *
+   * \param[in] pygoal_response_type An instance of the response message type to take.
+   * \return 2-tuple (sequence number, received response), or
+   * \return 2-tuple (None, None) if there is no response, or
+   * \return (None, None) if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:71]: @ahcorde won't it `raise` if there's a failure?

File: rclpy/src/rclpy/action_client.hpp [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+  /**
+   * Raises AttributeError if there is an issue parsing the pyfeedback_type.
+   * Raises RuntimeError on failure while taking a feedback message. Note, this does not include
+   * the case where there are no messages available.
+   * Raises RCLError if an error occurs in rcl
+   *
+   * \param[in] pyfeedback_type Instance of the feedback message type to take.
+   * \return Python message with all fields populated with received message, or
+   * \return None if there is nothing to take, or
+   * \return None if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:124]: @ahcorde won't it `raise` if there's a failure?

File: rclpy/src/rclpy/action_client.hpp [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+  /**
+   * Raises AttributeError if there is an issue parsing the pystatus_type.
+   * Raises RuntimeError on failure while taking a status message. Note, this does not include
+   * the case where there are no messages available.
+   * Raises RCLError if an error occurs in rcl
+   *
+   * \param[in] pystatus_type Instance of the status message type to take.
+   * \return Python message with all fields populated with received message, or
+   * \return None if there is nothing to take, or
+   * \return None if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:139]: @ahcorde won't it `raise` if there's a failure?

File: rclpy/src/rclpy/action_client.hpp [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+  /// Take an action result response.
+  /**
+   * Raises AttributeError if there is an issue parsing the pyresult_response_type.
+   * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
+   * Raises RCLError if an error occurs in rcl
+   *
+   * \param[in] pyresult_response_type An instance of the response message type to take.
+   * \return 2-tuple (sequence number, received response), or
+   * \return 2-tuple (None, None) if there is no response, or
+   * \return (None, None) if there is a failure.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:165]: @ahcorde won't it `raise` if there's a failure?

File: rclpy/src/rclpy/action_client.hpp [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+  bool
+  is_action_server_available();
+
+  /// Check if an action entity has any ready wait set entities.
+  /**
+   * This must be called after waiting on the wait set.
+   * Raises RuntimeError on failure.
+   *
+   * \param[in] pywait_set Capsule pointing to the wait set structure.
+   * \return A tuple of Bool representing the ready sub-entities.
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:197]: @ahcorde nit: 
```suggestion
   * \return A tuple of booleans representing the sub-entities ready:
```

**Review by hidmic — APPROVED** [review]
LGTM, comments are minor.

**Comment by ahcorde** [comment]
compiling up-to `rclpy`and testing `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14340/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9123/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12022/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14441/)

File: rclpy/src/rclpy/action_client.cpp [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+    "Take an action result response.")
+  .def(
+    "get_num_entities", &ActionClient::get_num_entities,
+    "Get the number of wait set entities that make up an action entity.")
+  .def(
+    "is_action_server_available", &ActionClient::is_action_server_available,
+    "Check if an action server is available for the given action client.")
+  .def(
+    "add_to_waitset", &ActionClient::add_to_waitset,
+    "Check if an action server is available for the given action client.")
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:322]: @ahcorde 
```suggestion
    "Add an action entity to a wait set.")
```
?

File: rclpy/src/rclpy/action_client.cpp [hunk 23]
```diff
@@ -0,0 +1,330 @@
...
+  .def(
+    "is_ready", &ActionClient::is_ready,
+    "Check if an action entity has any ready wait set entities.")
+  .def(
+    "take_status", &ActionClient::take_status,
+    "Add an action entitiy to a wait set.");
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.cpp:328]: @ahcorde 
```suggestion
    "Take an action status response.");
```

File: rclpy/src/rclpy/action_client.hpp [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+ * This class will create an action client for the given action name.
+ * This client will use the typesupport defined in the action module
+ * provided as pyaction_type to send messages.
+ */
+class ActionClient : public Destroyable, public std::enable_shared_from_this<ActionClient>
+{
+public:
+  /// Create an action client.
+  /*
+   * Raises ValueError if action name is invalid
```
Comment (hidmic) [inline rclpy/src/rclpy/action_client.hpp:41]: @ahcorde nit
```suggestion
   * Raises ValueError if action name is invalid.
```

File: rclpy/src/rclpy/utils.cpp [hunk 25]
```diff
@@ -81,6 +81,31 @@ create_from_py(py::object pymessage)
...
+  if (!convert) {
+    throw py::error_already_set();
+  }
+
+  if (!convert(pymessage.ptr(), message.get())) {
+    throw py::error_already_set();
+  }
+
+  return message;
+}
```
Comment (hidmic) [inline rclpy/src/rclpy/utils.cpp:107]: @ahcorde great addition !

**Review by hidmic — APPROVED** [review]

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14341/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9124/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12023/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14442/)

**Pull request merged**